### PR TITLE
Round up in edge cases around Vb undervalued

### PR DESCRIPTION
**Pull request by jubeira**
# Description

It's extremely unlikely if not impossible for the virtual balances to overflow in practice. But on paper they shouldn't, as the math is sound. Adjusting the rounding direction nudges the denominator of `Vu` to prevent it from underflowing in an extreme situation.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [x] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests have 100% code coverage
- [x] The base branch is either `main`, or there's a description of how to merge

## Issue Resolution

N/A

### CODE REVIEW

**Review by joaobrunoah — APPROVED** [review]
LGTM!

**Comment by jubeira** [comment]
Update: adjusted to `mulUp` in `virtualBalanceOvervalued` to prevent it from going to 0.
This also prevents an underflow if `Vo == 0` and `Ro == 0` (again, super extreme case that should never happen in practice, but this makes the math work as on paper, always).

**Review by EndymionJkb — APPROVED** [review]
Correct, and a strict improvement (modulo some comment nits, fixed in the follow-on), but the PoC can still fail in code before these rounding corrections. See #175 follow-on.

File: test/foundry/ReClammMath.t.sol [hunk 1]
```diff
@@ -624,4 +634,53 @@ contract ReClammMathTest is BaseReClammTest {
...
+        uint256 virtualBalanceUndervaluedDenominator = ((sqrtPriceRatio - FixedPoint.ONE).mulUp(
+            minVirtualBalanceOvervalued
+        ) - balanceScaledOvervalued);
+
+        assertGt(virtualBalanceUndervaluedDenominator, 0, "Virtual balance undervalued denominator should be positive");
+    }
+
+    /// @dev In the worst case scenario, Vo has to be greater than 0.
+    function testVirtualBalanceOvervalued__Fuzz(uint256 virtualBalanceOvervalued) public pure {
+        virtualBalanceOvervalued = bound(virtualBalanceOvervalued, 1, _MAX_TOKEN_BALANCE);
```
Comment (joaobrunoah) [inline test/foundry/ReClammMath.t.sol:675]: Not sure what is the value that the fuzz test brings here.

Vo will be the minimum possible when previous Vo is 1, dailyPriceShiftBase^duration is constant.

I'd test for previous Vo = 1 wei, for documentation purposes, and add reference to the function that is being tested here.

File: test/foundry/ReClammMath.t.sol [hunk 2]
```diff
@@ -624,4 +634,53 @@ contract ReClammMathTest is BaseReClammTest {
...
+        uint256 virtualBalanceUndervaluedDenominator = ((sqrtPriceRatio - FixedPoint.ONE).mulUp(
+            minVirtualBalanceOvervalued
+        ) - balanceScaledOvervalued);
+
+        assertGt(virtualBalanceUndervaluedDenominator, 0, "Virtual balance undervalued denominator should be positive");
+    }
+
+    /// @dev In the worst case scenario, Vu_denominator should be strictly positive to avoid underflow.
+    function testVirtualBalanceUndervaluedDenominatorUnderflowRoEqualThan0__Fuzz(uint256 sqrtPriceRatio) public view {
+        sqrtPriceRatio = bound(sqrtPriceRatio, _MIN_SQRT_PRICE_RATIO, _MAX_SQRT_PRICE_RATIO);
```
Comment (joaobrunoah) [inline test/foundry/ReClammMath.t.sol:659]: Here too, the worst case is the lowest sqrtPriceRatio. (mulUp is pretty straight forward when multiplying a constant).

I'd try it with the lowest possible sqrtPriceRatio only, don't need to call fuzz (unless you also want to test overflow).

File: test/foundry/ReClammMath.t.sol [hunk 3]
```diff
@@ -624,4 +634,53 @@ contract ReClammMathTest is BaseReClammTest {
...
     }
+
+    /// @dev In the worst case scenario, Vu_denominator should be strictly positive to avoid underflow.
+    function testVirtualBalanceUndervaluedDenominatorUnderflowRoGreaterThan0__Fuzz(
+        uint256 sqrtPriceRatio,
+        uint256 balanceScaledOvervalued
+    ) public view {
+        sqrtPriceRatio = bound(sqrtPriceRatio, _MIN_SQRT_PRICE_RATIO, _MAX_SQRT_PRICE_RATIO);
+        // Ro can't be 0 in this case
+        balanceScaledOvervalued = bound(balanceScaledOvervalued, 1, _MAX_TOKEN_BALANCE);
```
Comment (joaobrunoah) [inline test/foundry/ReClammMath.t.sol:645]: In here fuzz test makes more sense, with two parameters it gets trickier to evaluate.

**Review by joaobrunoah — APPROVED** [review]
LGTM, replacing rounding on VBs to divUp and mulUp makes sense, it's only a shift of 1 wei up. (Swap fees affect the pool concentration much more than this).

I just did some observations in the tests.
